aws/stscreds/web_identity_provider.go

```diff
@@ -0,0 +1,103 @@
+package stscreds
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/aws/aws-sdk-go-v2/service/sts/stsiface"
+)
+
+const (
+	// ErrCodeWebIdentity will be used as an error code when constructing
+	// a new error to be returned during session creation or retrieval.
+	ErrCodeWebIdentity = "WebIdentityErr"
+
+	// WebIdentityProviderName is the web identity provider name
+	WebIdentityProviderName = "WebIdentityCredentials"
+)
+
+// WebIdentityRoleProvider is used to retrieve credentials using
+// an OIDC token.
+type WebIdentityRoleProvider struct {
+	aws.SafeCredentialsProvider
+
+	client stsiface.ClientAPI
+
+	tokenFilePath   string
+	roleARN         string
+	roleSessionName string
+
+	options WebIdentityRoleProviderOptions
+}
+
+// WebIdentityRoleProviderOptions is a structure of configurable options for WebIdentityRoleProvider
+type WebIdentityRoleProviderOptions struct {
+	ExpiryWindow time.Duration
+}
+
+// NewWebIdentityRoleProvider will return a new WebIdentityRoleProvider with the
+// provided stsiface.STSAPI
+func NewWebIdentityRoleProvider(svc stsiface.ClientAPI, roleARN, roleSessionName, path string, options ...func(*WebIdentityRoleProviderOptions)) *WebIdentityRoleProvider {
```

Comment:
What do you think about changing the `path` parameter to a interface that has a method to get the JWT bytes. 
Minor difference, but it decouples where the JWT bytes come from outside of the web identity role provider. This also has a small benefit for tests, that a actual JWT file isn't needed for unit testing this provider.

```go
type JWTBytes interface {
    GetJWTBytes() ([]byte, error)
}

type JWTFilename string
func (f *FileBytes) GetJWTBytes() ([]byte, error) { /* ... */ }
```

Usage:
```go
stscreds.NewWebIdentityRoleProvider(client, myARN, "sessionName", stscreds.JWTFilename("myfile.jwt"))
```